### DISCUSSION TITLE
Polymorphism?

### DISCUSSION THREAD

**Discussion by liamjcooper** [area17/twill#870] [Ideas]
Just putting this idea out here, would polymorphism for the module features be a welcome change for Twill?

Imagine you create any old module like normal, but instead of the migration creating possibly 3 extra tables for features such as post_revisions, post_translations and post_slugs, why not add the polymorphic relationship into the generated model via the trait?

For example, Twill installs itself into a fresh project with the tables it does now, with 3 extra ones: slugs, translations and revisions (these could be prefixed with twill_ if necessary). Each table has the same schema as what you would expect with how Twill creates these tables normally via modules, with the 2 extra columns as documented in Laravel i.e. "translatable_type" and "translatable_id" for the translations table. The "type" has the resolvable model name and the id is the id of the record. Maybe looking forward we should have an indexable "twill_id" for every module, as the developer might have configured a different primary key for the table rather than a regular unsignedBigInteger.

[More info on how polymorphism works in Laravel here.](https://laravel.com/docs/8.x/eloquent-relationships#polymorphic-relationships)

This negates the need for up to 3 extra tables per module (so cleaner migrations and a neater table view) and we wouldn't require the extra feature directories in the Models directory e.g. app/Models/Slugs as the models for the features would essentially be standalone and the module trait would set up the relationship for it.

Is this something worth spending time on, or is there a particular reason this hasn't been implemented? This would be a breaking change in my eyes, so I'd rather have a conversation about it first.

Many thanks.

**Comment by daun** [comment]
Revisions and slugs I could see working that way, but the translation tables have columns that are specific to each module, right? You'd have to put the translations into a json column which would mess with performance and query logic.

**Reply by liamjcooper** [reply]
> Revisions and slugs I could see working that way, but the translation tables have columns that are specific to each module, right? You'd have to put the translations into a json column which would mess with performance and query logic.

Yeah, good point. That's obviously a bit more tricky.

I would suggest maybe having a base translations table that contains the locale, active, timestamps, ids and the polymorphic columns and then an intermediary table (like a pivot table?) which has the columns unique to the module it belongs to, i.e. posts_translations. But I dunno, maybe that's over-engineering it? At the very least it would drop the `createDefaultTranslationsTableFields()` call from the module migration as the developer would just be creating the columns they actually want translating. Then the model relationship fills in the rest?

**Reply by daun** [reply]
I feel like you'd be optimising the wrong things here. Removing boilerplate from migration files or reducing the number of tables is secondary. What you want is high query performance and few moving parts. With those canonical translations tables you have efficient indexes, immediately obvious project structure and as few sql queries as possible.

**Reply by liamjcooper** [reply]
Fair enough, if performance is more a priority here that's fine. Would this idea still make sense for the slugs and revisions features?

**Reply by ifox** [reply]
I'll echo @daun's points here. We do heavily use polymorphism already for medias and blocks, but it doesn't necessarily makes sense for everything. Slugs tables can also have columns that are specific to a model. Revisions might be a candidate, though the fact that they are quite heavy makes me feel like it's a good thing that they are not all in the same table. I think we would gain more value by refactoring migrations helpers to simplify the process, maybe? Ultimately in the future we want to generate more code and free developers up from managing migrations manually, but we will never want to prevent hooking up existing tables/models into Twill.

**Reply by liamjcooper** [reply]
Gotcha, loud and clear. Thanks guys
